### ISSUE TITLE
Clipmenud not responding to systemd SIGTERM.

### ISSUE THREAD

**Issue by Filip62** [cdown/clipmenu#139]
OS: Arch Linux
Clipmenu: 6.1.0

My system started taking a long time to shutdown because systemd was running a stop job for 1m30s for a program of my user session. After quite a bit of debugging I found that clipmenud was not exiting on the SIGTERM that systemd sent.

```bash
# Kill all background processes on exit
trap 'trap - TERM; kill -- -$$' INT TERM EXIT
```

After I changed it to this: (removed the dash in front of $$)
```bash
# Kill all background processes on exit
trap 'trap - TERM; kill -- $$' INT TERM EXIT
```

My system shut down properly.

Though while I was debugging this I noticed that while it does terminate the clipmenud process, the clipnotify process clipmenud spawned does not get terminated, so if I run clipmenud again, it complains that another clipmenud is running, but it is not, just the clipnotify process that was spawned by the first clipmenud and did not get terminated.

**Comment by cdown** [comment]
That's not a bug, `-$$` means "terminate all pids with PGID `$$`". Which processes are still running after the SIGTERM?

**Comment by Filip62** [comment]
Right after boot and login:
![image](https://user-images.githubusercontent.com/58399088/95763624-93001b80-0caf-11eb-846e-0ff167cbe02e.png)

Right after I send SIGTERM to bash /usr/bin/clipmenud:
![image](https://user-images.githubusercontent.com/58399088/95763834-e4a8a600-0caf-11eb-94d9-e7b398ba2268.png)

Log:
`/usr/bin/clipmenud: line 1: kill: (-680) - No such process`

It looks like the dash just gets prepended.
Just to be clear, I do not know bash scripting. I am sorry I assumed it was a bug.

**Comment by marc1984** [comment]
Experiencing the same issue in my Arch Linux installation. Had to do a little bit of investigation to find out in the end that the process causing the shutdown/reboot delays was clipmenud. In my case though I'm starting clipmenud using the autostart patch from DWM. 

I tried to troubleshoot the error by creating my own systemd service unit, and although systemd is then able to kill clipmenud normally, it seems that clipnotify never starts and thus clipmenu cannot copy anything.

**Comment by Filip62** [comment]
@marc1984 There is a systemd unit written by the author.
[Clipmenud servive file](https://www.github.com/cdown/clipmenu/tree/develop/init%2Fclipmenud.service)

**Comment by cdown** [comment]
I'm moderately confused that this pid isn't the same as the pgrp, but what's actually at pid 628?

Either way, we can manually check the pgrp when shutting down, instead.

**Comment by cdown** [comment]
And wait, why are there two clipnotify daemons? That doesn't make much sense. clipnotify is explicitly awaited with waitpid, so it shouldn't be possible for two to run at once.

**Comment by marc1984** [comment]
Thanks @Filip62, I was looking under /etc/systemd/system and not in the user space. So turns out that using systemd to start up the service and it gets properly killed with SIGTERM on shutdown (no delays). The issue though is that when starting clipmenud using the packaged user service unit clipmenu does not work, cannot copy text selections. Am I missing something or is it a bug? 
![systemctl_clipmenud_status](https://user-images.githubusercontent.com/9002794/95915688-d3db5b80-0da7-11eb-9166-592267befcba.png)

**Comment by cdown** [comment]
This looks like a totally separate issue, but what does `sed -nz /DISPLAY/p /proc/1108/environ` say? Is that right value?

**Comment by cdown** [comment]
I'm still not entirely sure how it happens that we're not the process group leader in an interactive session, but it's true that if there's a wrapper script or something else in the way, it might not work out.

@Filip62, can you please try 8d29b2aaf841d5346dbcc2c5e18cde4f662807f1 and let me know if it fixes your problem?

**Comment by Filip62** [comment]
If I use the provided systemd clipmenud service, everything works just as expected. Only clipmenud and clipnotify have the same PGRP and SIGTERM terminates both, no shutdown problems.

If I use [8d29b2a](https://github.com/cdown/clipmenu/commit/8d29b2aaf841d5346dbcc2c5e18cde4f662807f1), then also everything works as expected.

Plese wait a minute for an edit after reboot.

**Comment by cdown** [comment]
Yeah. 6.2.0 is now out, although this shouldn't affect many people hopefully. Let me know if there are any problems :-)

**Comment by marc1984** [comment]
> This looks like a totally separate issue, but what does sed -nz /DISPLAY/p /proc/1108/environ say? Is that right value?
DISPLAY=:0%

**Comment by cdown** [comment]
Hmm, looks like a different issue then. Feel free to file if it's still happening for you, there are a couple of fixes for stuff which went out in 6.2.0 :-)
